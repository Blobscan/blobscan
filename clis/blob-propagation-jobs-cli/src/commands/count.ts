import commandLineArgs from "command-line-args";
import commandLineUsage from "command-line-usage";

import type { QueueHumanName } from "../Context";
import { context } from "../context-instance";
import type { Command } from "../types";
import { helpOptionDef, queuesOptionDef } from "../utils";

const countCommandOptDefs: commandLineArgs.OptionDefinition[] = [
  helpOptionDef,
  {
    ...queuesOptionDef,
    description: `Queue to get the jobs count of. ${queuesOptionDef.description}`,
  },
];

export const countCommandUsage = commandLineUsage([
  {
    header: "Count Command",
    content: "Count propagation jobs.",
  },
  {
    header: "Options",
    optionList: countCommandOptDefs,
  },
]);

export const count: Command = async function (argv) {
  const { help, queue: queueNames } = commandLineArgs(countCommandOptDefs, {
    argv,
  }) as {
    help?: boolean;
    queue?: QueueHumanName[];
  };

  if (help) {
    console.log(countCommandUsage);

    return;
  }

  const queues = queueNames
    ? context.getQueues(queueNames)
    : context.getAllQueues();

  const jobCounts = await Promise.all(
    queues.map(async (q) => ({
      queueName: q.name,
      count: await q.getJobCounts(),
    }))
  );

  console.log(jobCounts);
};
