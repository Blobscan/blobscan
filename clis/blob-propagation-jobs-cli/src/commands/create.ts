import commandLineArgs from "command-line-args";
import commandLineUsage from "command-line-usage";

import {
  FINALIZER_WORKER_NAME,
  createBlobPropagationFlowJob,
} from "@blobscan/blob-propagator";
import { prisma } from "@blobscan/db";

import { context } from "../context-instance";
import { env } from "../env";
import type { Command } from "../utils";
import {
  blobHashOptionDef,
  datePeriodOptionDefs,
  helpOptionDef,
  normalizeBlobHashes,
  normalizeDate,
  normalizeStorageQueueName,
} from "../utils";

// TODO: Need to convert it to number explicity due to ci tests failing
// with the following error:
const PRISMA_BATCH_OPERATIONS_MAX_SIZE = parseInt(
  env.PRISMA_BATCH_OPERATIONS_MAX_SIZE.toString()
);
const createCommandOptDefs: commandLineArgs.OptionDefinition[] = [
  helpOptionDef,
  {
    ...blobHashOptionDef,
    description: "Blob hash of the blobs to create jobs for.",
  },
  {
    name: "storage",
    alias: "s",
    typeLabel: "{underline storage}",
    description:
      "Storage used to propagate the selected blobs. Valid values are {italic google}, {italic postgres} or {italic swarm}.",
    multiple: true,
    type: String,
  },
  {
    ...datePeriodOptionDefs.from,
    description: "Date from which to retrieve blobs to create jobs for.",
  },
  {
    ...datePeriodOptionDefs.to,
    description: "Date to which to retrieve blobs to create jobs for.",
  },
];

export const createCommandUsage = commandLineUsage([
  {
    header: "Create Command",
    content: "Create propagation jobs for blobs.",
  },
  {
    header: "Options",
    optionList: createCommandOptDefs,
  },
]);

async function createBlobPropagationJobsInBatches(
  storageQueueNames: string[],
  blobHashesBatchFetcher: (cursorId?: string) => Promise<{
    blobHashes: string[];
    nextCursorId?: string;
  }>
) {
  let cursorId: string | undefined;

  do {
    const { blobHashes, nextCursorId } = await blobHashesBatchFetcher(cursorId);

    const jobs = blobHashes.map((blobHash) =>
      createBlobPropagationFlowJob(
        FINALIZER_WORKER_NAME,
        storageQueueNames,
        blobHash
      )
    );

    if (jobs.length) {
      await context.getPropagatorFlowProducer().addBulk(jobs);
    }

    cursorId = nextCursorId;
  } while (cursorId);
}

export const create: Command = async function (argv) {
  const {
    blobHash: rawBlobHashes,
    help,
    storage: rawStorageNames,
    from: rawFrom,
    to: rawTo,
  } = commandLineArgs(createCommandOptDefs, {
    argv,
  }) as {
    blobHash?: string[];
    help?: boolean;
    storage?: string[];
    from?: string;
    to?: string;
  };

  if (help) {
    console.log(createCommandUsage);

    return;
  }

  const from = rawFrom ? normalizeDate(rawFrom) : undefined;
  const to = rawTo ? normalizeDate(rawTo) : undefined;
  const argStorageNames = rawStorageNames?.map((rawName) =>
    normalizeStorageQueueName(rawName)
  );
  const storageQueues = argStorageNames
    ? context.getQueuesOrThrow(argStorageNames)
    : context.getAllStorageQueues();
  const storageQueueNames = storageQueues.map((q) => q.name);
  let blobHashes: string[];

  if (rawBlobHashes?.length) {
    blobHashes = normalizeBlobHashes(rawBlobHashes);

    const dbBlobs = await prisma.blob.findMany({
      where: {
        versionedHash: {
          in: blobHashes,
        },
      },
    });

    if (dbBlobs.length !== blobHashes.length) {
      const notFoundBlobHashes = blobHashes.filter(
        (hash) => !dbBlobs.find((b) => b.versionedHash === hash)
      );

      throw new Error(
        `Could not find blobs with the following hashes: ${notFoundBlobHashes.join(
          ", "
        )}`
      );
    }

    const jobs = blobHashes.map((blobHash) =>
      createBlobPropagationFlowJob(
        FINALIZER_WORKER_NAME,
        storageQueueNames,
        blobHash
      )
    );

    await context.getPropagatorFlowProducer().addBulk(jobs);
  } else if (from || to) {
    await createBlobPropagationJobsInBatches(
      storageQueueNames,
      async (cursorId) => {
        const dbBlocks = await prisma.block.findMany({
          take: PRISMA_BATCH_OPERATIONS_MAX_SIZE,
          skip: cursorId ? 1 : undefined,
          cursor: cursorId
            ? {
                hash: cursorId,
              }
            : undefined,
          select: {
            hash: true,
            transactions: {
              select: {
                blobs: {
                  select: {
                    blobHash: true,
                  },
                },
              },
            },
          },
          where: {
            timestamp: {
              gte: from,
              lte: to,
            },
          },
        });

        const blobHashes = [
          ...new Set(
            dbBlocks
              .map((b) =>
                b.transactions.map((t) => t.blobs.map((bl) => bl.blobHash))
              )
              .flat(2)
          ),
        ];

        return {
          blobHashes,
          nextCursorId: dbBlocks[dbBlocks.length - 1]?.hash,
        };
      }
    );
  } else {
    await createBlobPropagationJobsInBatches(
      storageQueueNames,
      async (cursorId) => {
        const dbBlobs = await prisma.blob.findMany({
          take: PRISMA_BATCH_OPERATIONS_MAX_SIZE,
          cursor: cursorId
            ? {
                commitment: cursorId,
              }
            : undefined,
          skip: cursorId ? 1 : undefined,
        });
        const blobHashes = dbBlobs.map((b) => b.versionedHash);

        return {
          blobHashes,
          nextCursorId: dbBlobs[dbBlobs.length - 1]?.commitment,
        };
      }
    );
  }
};
