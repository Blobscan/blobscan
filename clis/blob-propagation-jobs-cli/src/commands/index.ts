export * from "./remove";
export * from "./retry";
