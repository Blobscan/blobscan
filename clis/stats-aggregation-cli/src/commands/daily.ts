import commandLineArgs from "command-line-args";
import commandLineUsage from "command-line-usage";

import dayjs from "@blobscan/dayjs";
import type { DatePeriod, RawDatePeriod } from "@blobscan/db";
import { normalizeDailyDatePeriod, prisma } from "@blobscan/db";

import { CommandError } from "../error";
import type { Command, Entity } from "../types";
import {
  ALL_ENTITIES,
  commandLog,
  deleteOptionDef,
  helpOptionDef,
} from "../utils";

type Operation = "populate" | "deleteMany" | "deleteAll";

type OperationResult = { entity: Entity; affectedRows: number };
class DailyCommandError extends CommandError {
  constructor(operation: Operation, error: Error | string) {
    const formattedOperation =
      operation === "deleteMany" ? "deletion" : "aggregation";
    const message = error instanceof Error ? error.message : error;

    super(message, {
      command: "daily",
      operation: formattedOperation,
    });
  }
}

const dailyCommandOptDefs: commandLineUsage.OptionDefinition[] = [
  deleteOptionDef,
  helpOptionDef,
  {
    name: "entity",
    alias: "e",
    typeLabel: "{underline type}",
    description:
      "Entity type to aggregate. Valid values are {italic blob}, {italic block} or {italic tx}.",
    type: String,
    multiple: true,
  },
  {
    name: "from",
    alias: "f",
    typeLabel: "{underline date}",
    description: "Start date in ISO 8601 format.",
    type: String,
  },
  {
    name: "to",
    alias: "t",
    typeLabel: "{underline date}",
    description: "End date in ISO 8601 format.",
    type: String,
  },
];

export const dailyCommandUsage = commandLineUsage([
  {
    header: "Daily Command",
    content: "Aggregate daily stats.",
  },
  {
    header: "Options",
    optionList: dailyCommandOptDefs,
  },
]);

function buildCommandResultMsg(
  operation: Operation,
  datePeriod: DatePeriod,
  opResults: OperationResult[]
) {
  const formattedOperation =
    operation === "deleteMany" ? "deleted" : "calculated";
  const { from, to } = datePeriod || {};
  const formattedFrom = from ? dayjs(from).format("YYYY/MM/DD") : undefined;
  const formattedTo = to ? dayjs(to).format("YYYY/MM/DD") : undefined;
  let period = "";

  if (!from && to) {
    period = `up to ${formattedTo}`;
  } else if (from && !to) {
    period = `from ${formattedFrom}`;
  } else if (from && to) {
    period = `from ${formattedFrom} to ${formattedTo}`;
  }

  const formattedResults =
    opResults
      ?.map((r) => `${r.entity} (${r.affectedRows} rows affected)`)
      .join(", ") || "";

  return `stats ${formattedOperation} successfully ${period} for entities: ${formattedResults}`;
}

async function performDailyStatsOperation(
  entity: Entity,
  operation: Operation,
  datePeriod?: DatePeriod
) {
  let dailyStatsFn;
  const { from, to } = datePeriod || {};
  const isDateProvided = from || to;
  const operation_: Operation =
    operation === "deleteMany" && !isDateProvided ? "deleteAll" : operation;
  const operationParam =
    operation_ === "populate"
      ? datePeriod
      : isDateProvided
      ? {
          where: {
            day: {
              ...(from ? { gte: from } : {}),
              ...(to ? { lte: to } : {}),
            },
          },
        }
      : undefined;

  switch (entity) {
    case "blob":
      dailyStatsFn = prisma.blobDailyStats[operation_];
      break;
    case "block":
      dailyStatsFn = prisma.blockDailyStats[operation_];
      break;
    case "tx":
      dailyStatsFn = prisma.transactionDailyStats[operation_];
      break;
  }

  let result;

  try {
    result = await dailyStatsFn(operationParam).then((res) => {
      return typeof res === "number" ? res : res.count;
    });
  } catch (err) {
    const err_ = err as Error;

    throw new DailyCommandError(
      operation,
      `Failed to perform operation for entity ${entity}: ${err_.message}`
    );
  }

  return result;
}

const daily: Command = async function daily(argv) {
  const {
    from: rawFrom,
    to: rawTo,
    entity: entities,
    // `delete` is a reserved keyword
    delete: delete_,
    help,
  } = commandLineArgs(dailyCommandOptDefs, {
    argv,
  }) as {
    help: boolean;
    delete?: boolean;
    from?: string;
    to?: string;
    entity?: Entity[];
  };

  if (help) {
    console.log(dailyCommandUsage);

    return;
  }

  const operation: Operation = delete_ ? "deleteMany" : "populate";
  const rawDatePeriod: RawDatePeriod = { from: rawFrom, to: rawTo };
  let datePeriod: DatePeriod;
  try {
    datePeriod = normalizeDailyDatePeriod(rawDatePeriod);
  } catch (err) {
    const err_ = err as Error;

    throw new DailyCommandError(operation, err_);
  }
  const selectedEntities = entities?.length ? entities : ALL_ENTITIES;

  const operationResults: OperationResult[] = await Promise.all(
    selectedEntities.map((e) =>
      performDailyStatsOperation(e, operation, datePeriod).then(
        (affectedRows) => ({
          entity: e,
          affectedRows,
        })
      )
    )
  );

  const msg = buildCommandResultMsg(operation, datePeriod, operationResults);
  commandLog("info", "daily", operation, msg);
};

export { daily };
