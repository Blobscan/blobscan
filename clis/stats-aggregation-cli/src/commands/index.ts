export * from "./daily";
export * from "./overall";
