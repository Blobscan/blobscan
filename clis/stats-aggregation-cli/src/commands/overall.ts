import commandLineArgs from "command-line-args";
import commandLineUsage from "command-line-usage";

import type { BlockNumberRange, Prisma } from "@blobscan/db";
import { prisma } from "@blobscan/db";

import { CommandError } from "../error";
import type { Command } from "../types";
import { commandLog, deleteOptionDef, helpOptionDef } from "../utils";

type BlockId = number | "latest" | "finalized";

const PRISMA_BLOCKS_BATCH_SIZE = 2_000_000;

class OverallCommandError extends CommandError {
  constructor(operation: string, error: Error | string) {
    super(error instanceof Error ? error.message : error, {
      command: "overall",
      operation,
    });
  }
}

const overallCommandOptDefs: commandLineUsage.OptionDefinition[] = [
  deleteOptionDef,
  helpOptionDef,
  {
    name: "to",
    alias: "t",
    typeLabel: "{underline block-id}",
    description: `Block identifier up to which to aggregate data. It can be a block number, "latest" for the last indexed block or "finalized"  for the chain's most recent finalized block. It defaults to "finalized"`,
    defaultValue: "finalized",
    type: String,
  },
  {
    name: "batchSize",
    alias: "s",
    typeLabel: "{underline size}",
    description: `Number of blocks to process in a single batch. It defaults to ${PRISMA_BLOCKS_BATCH_SIZE}`,
    defaultValue: PRISMA_BLOCKS_BATCH_SIZE,
    type: Number,
  },
];

export const overallCommandUsage = commandLineUsage([
  {
    header: "Overall Command",
    content: "Aggregate overall stats.",
  },
  {
    header: "Options",
    optionList: overallCommandOptDefs,
  },
]);

async function resolveBlockId(blockId: BlockId): Promise<number> {
  try {
    if (typeof blockId === "number" || !isNaN(Number(blockId))) {
      return Number(blockId);
    } else if (blockId === "finalized") {
      const syncState = await prisma.blockchainSyncState.findUnique({
        select: {
          lastFinalizedBlock: true,
        },
        where: {
          id: 1,
        },
      });

      return syncState?.lastFinalizedBlock ?? 0;
    } else {
      return (await prisma.block.findLatest().then((b) => b?.number)) ?? 0;
    }
  } catch (err) {
    const err_ = err as Error;

    throw new OverallCommandError("aggregation", err_);
  }
}

async function deleteOverallStats() {
  const newBlockchainSyncState: Partial<Prisma.BlockchainSyncStateGroupByOutputType> =
    {
      lastAggregatedBlock: null,
    };

  await prisma.$transaction([
    prisma.blobOverallStats.deleteMany(),
    prisma.blockOverallStats.deleteMany(),
    prisma.transactionOverallStats.deleteMany(),
    prisma.blockchainSyncState.upsert({
      create: newBlockchainSyncState,
      update: newBlockchainSyncState,
      where: {
        id: 1,
      },
    }),
  ]);
}

async function incrementOverallStats(
  blockRange: BlockNumberRange,
  {
    batchSize = PRISMA_BLOCKS_BATCH_SIZE,
  }: {
    batchSize?: number;
  }
) {
  const { from, to } = blockRange;
  const unprocessedBlocks = to - from + 1;
  const batches = Math.ceil(unprocessedBlocks / batchSize);

  for (let i = 0; i < batches; i++) {
    const batchFrom = from + i * batchSize;
    const batchTo = Math.min(batchFrom + batchSize - 1, to);
    const blockRange: BlockNumberRange = { from: batchFrom, to: batchTo };
    const newBlockchainSyncState: Partial<Prisma.BlockchainSyncStateGroupByOutputType> =
      {
        lastAggregatedBlock: batchTo,
      };

    await prisma.$transaction([
      prisma.blockOverallStats.increment(blockRange),
      prisma.transactionOverallStats.increment(blockRange),
      prisma.blobOverallStats.increment(blockRange),
      prisma.blockchainSyncState.upsert({
        create: newBlockchainSyncState,
        update: newBlockchainSyncState,
        where: {
          id: 1,
        },
      }),
    ]);

    if (batches > 1) {
      const formattedBlockRange = `${blockRange.from.toLocaleString()}-${blockRange.to.toLocaleString()}`;
      const batch = batches > 1 ? `(batch ${i + 1}/${batches})` : "";

      commandLog(
        "debug",
        "overall",
        "increment",
        `stats calculated for block range ${formattedBlockRange} (${batch})`
      );
    }
  }
}

const overall: Command = async function (argv) {
  const {
    batchSize,
    to,
    delete: delete_,
    help,
  } = commandLineArgs(overallCommandOptDefs, {
    argv,
  }) as {
    batchSize: number;
    to?: string;
    delete?: boolean;
    help: boolean;
  };

  if (help) {
    console.log(overallCommandUsage);

    return;
  }

  if (delete_) {
    await deleteOverallStats();

    commandLog(
      "info",
      "overall",
      "deleteMany",
      `All overall stats removed successfully`
    );

    return;
  }

  let targetBlockId: BlockId;

  if (!isNaN(Number(to))) {
    targetBlockId = Number(to);
  } else if (to === "latest" || to === "finalized") {
    targetBlockId = to;
  } else {
    throw new OverallCommandError(
      "aggregation",
      `Invalid \`to\` flag value. Expected a block number, "latest" or "finalized" but got ${to}`
    );
  }
  let targetBlockNumber = await resolveBlockId(targetBlockId);
  const [lastSyncedFinalizedBlock, latestIndexedBlock] = await Promise.all([
    prisma.blockchainSyncState
      .findUnique({
        where: {
          id: 1,
        },
      })
      .then((state) => state?.lastAggregatedBlock ?? 0),
    prisma.block.findLatest().then((b) => b?.number),
  ]);

  // If we haven't indexed any blocks yet, don't do anything
  if (!latestIndexedBlock) {
    commandLog(
      "info",
      "overall",
      "increment",
      "Skipping as there as no blocks have been indexed yet"
    );

    return;
  }

  const fromBlockNumber = lastSyncedFinalizedBlock + 1;

  // Process only finalized blocks that were indexed
  if (targetBlockNumber > latestIndexedBlock) {
    targetBlockNumber = latestIndexedBlock;
  }

  if (fromBlockNumber > targetBlockNumber) {
    commandLog(
      "info",
      "overall",
      "increment",
      `Skipping as there is no new finalized blocks (Last processed finalized block: ${lastSyncedFinalizedBlock.toLocaleString()})`
    );

    return;
  }

  await incrementOverallStats(
    { from: fromBlockNumber, to: targetBlockNumber },
    { batchSize }
  );

  commandLog(
    "info",
    "overall",
    "increment",
    `All stats incremented successfully from block ${fromBlockNumber.toLocaleString()} to ${targetBlockNumber.toLocaleString()}`
  );

  return;
};

export { overall };
