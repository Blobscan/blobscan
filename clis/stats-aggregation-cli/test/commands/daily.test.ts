import {
  afterAll,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from "vitest";

import { prisma } from "@blobscan/db";
import { fixtures } from "@blobscan/test";

import { daily, dailyCommandUsage } from "../../src/commands/daily";
import {
  getDailyStatsByDateRange,
  runHelpArgTests,
  toDailyFormat,
} from "../helpers";

function sortByDay(a: { day: string | Date }, b: { day: string | Date }) {
  return new Date(a.day).getTime() - new Date(b.day).getTime();
}

function getAllDailyStats() {
  return Promise.all([
    prisma.blobDailyStats.findMany(),
    prisma.blockDailyStats.findMany(),
    prisma.transactionDailyStats.findMany(),
  ]).then((allDailyStats) =>
    allDailyStats.map((stats) => stats.sort(sortByDay))
  );
}

const DATE_SORTED_BLOCKS = fixtures.blocks.sort((a, b) => {
  return new Date(a.insertedAt).getTime() - new Date(b.insertedAt).getTime();
});

describe("Daily command", () => {
  runHelpArgTests(daily, dailyCommandUsage);

  beforeAll(() => {
    // Silence console.log
    vi.spyOn(console, "log").mockImplementation(() => void {});
  });

  afterAll(() => {
    vi.clearAllMocks();
  });

  describe("when aggregating daily stats", () => {
    it("should aggregate all entities' data when no entity is provided", async () => {
      await daily();

      const [blobDailyStats, blockDailyStats, transactionDailyStats] =
        await getAllDailyStats();

      expect(blobDailyStats, "Blob daily stats mismatch").toMatchSnapshot();
      expect(blockDailyStats, "Block daily stats mismatch").toMatchSnapshot();
      expect(
        transactionDailyStats,
        "Transaction daily stats mismatch"
      ).toMatchSnapshot();
    });

    it("should aggregate blob data when the blob entity is provided", async () => {
      await daily(["--entity", "blob"]);

      const [blobDailyStats, blockDailyStats, transactionDailyStats] =
        await getAllDailyStats();

      expect(blobDailyStats, "Blob daily stats mismatch").toMatchSnapshot();
      expect(blockDailyStats, "Block daily stats should be empty").toEqual([]);
      expect(
        transactionDailyStats,
        "Transaction daily stats should be empty"
      ).toEqual([]);
    });

    describe("when specifying a date period", () => {
      it("should aggregate data from the provided period", async () => {
        const fromDate = "2023-05-05";
        const toDate = "2023-08-30";

        await daily(["--from", fromDate, "--to", toDate, "--entity", "blob"]);

        const statsOutsideRange = getDailyStatsByDateRange(
          await prisma.blobDailyStats.findMany(),
          {
            fromDate,
            toDate,
            invertRange: true,
          }
        );

        expect(statsOutsideRange).toEqual([]);
      });

      it("should aggregate all data when no period is provided", async () => {
        const expectedEarliestDailyStat = DATE_SORTED_BLOCKS[0];
        const expectedEarliestDate = expectedEarliestDailyStat
          ? toDailyFormat(expectedEarliestDailyStat.insertedAt)
          : null;
        const expectedLatestDailyStat =
          DATE_SORTED_BLOCKS[DATE_SORTED_BLOCKS.length - 1];
        const expectedLatestDate = expectedLatestDailyStat
          ? toDailyFormat(expectedLatestDailyStat.insertedAt)
          : null;

        await daily(["--entity", "block"]);

        const blobDailyStats = (await prisma.blockDailyStats.findMany()).sort(
          (a, b) => a.day.getTime() - b.day.getTime()
        );

        const earliestDailyStat = blobDailyStats[0];
        const earliestDate = earliestDailyStat
          ? toDailyFormat(earliestDailyStat.day)
          : null;
        const latestDailyStat = blobDailyStats[blobDailyStats.length - 1];
        const latestDate = latestDailyStat
          ? toDailyFormat(latestDailyStat.day)
          : null;

        expect(earliestDate, "Earliest date mismatch").toBe(
          expectedEarliestDate
        );
        expect(latestDate, "Latest date mismatch").toBe(expectedLatestDate);
      });

      it("should aggregate data until the latest date when no end date is provided", async () => {
        const fromDate = "2023-05-05";

        await daily(["--from", fromDate, "--entity", "block"]);

        const statsOutsideRange = getDailyStatsByDateRange(
          await prisma.blockDailyStats.findMany(),
          {
            fromDate,
            invertRange: true,
          }
        );

        expect(statsOutsideRange).toEqual([]);
      });

      it("should aggregate data from the earliest date when no start date is provided", async () => {
        const toDate = "2023-08-05";

        await daily(["--to", toDate, "--entity", "block"]);

        const statsOutsideRange = getDailyStatsByDateRange(
          await prisma.blockDailyStats.findMany(),
          {
            toDate,
            invertRange: true,
          }
        );

        expect(statsOutsideRange).toEqual([]);
      });

      it("should fail when starting date is after ending date", async () => {
        await expect(
          daily(["--from", "2021-01-02", "--to", "2021-01-01"])
        ).rejects.toThrowErrorMatchingInlineSnapshot(
          '"Daily stats aggregation failed: Invalid date period. Start date is after end date"'
        );
      });
    });
  });

  describe("when deleting daily stats", () => {
    beforeEach(async () => {
      await daily();
    });

    it("should delete all daily stats when no entity is provided", async () => {
      await daily(["--delete"]);

      const [blobDailyStats, blockDailyStats, transactionDailyStats] =
        await getAllDailyStats();

      expect(blobDailyStats, "Blob daily stats should be empty").toEqual([]);
      expect(blockDailyStats, "Block daily stats should be empty").toEqual([]);
      expect(
        transactionDailyStats,
        "Transaction daily stats should be empty"
      ).toEqual([]);
    });

    it("should delete blob daily stats when the blob entity is provided", async () => {
      const [_, beforeBlockDailyStats, beforeTransactionDailyStats] =
        await getAllDailyStats();

      await daily(["--delete", "--entity", "blob"]);

      const [blobDailyStats, blockDailyStats, transactionDailyStats] =
        await getAllDailyStats();

      expect(blobDailyStats, "Blob daily stats should be empty").toEqual([]);
      expect(blockDailyStats, "Block daily stats mismatch").toEqual(
        beforeBlockDailyStats
      );
      expect(transactionDailyStats, "Transaction daily stats mismatch").toEqual(
        beforeTransactionDailyStats
      );
    });

    describe("when specifying a date period", () => {
      it("should delete data from a given period", async () => {
        const fromDate = "2023-05-05";
        const toDate = "2023-08-30";

        await daily([
          "--from",
          fromDate,
          "--to",
          toDate,
          "--delete",
          "--entity",
          "blob",
        ]);

        const blobDailyStats = getDailyStatsByDateRange(
          await prisma.blobDailyStats.findMany(),
          {
            fromDate,
            toDate,
          }
        );

        expect(blobDailyStats).toEqual([]);
      });

      it("should delete all data when no period is provided", async () => {
        await daily(["--delete", "--entity", "block"]);

        const blobDailyStats = await prisma.blockDailyStats.findMany();

        expect(blobDailyStats, "Block daily stats should be empty").toEqual([]);
      });

      it("should delete data until the latest date when no end date is provided", async () => {
        const fromDate = "2023-05-05";

        await daily(["--from", fromDate, "--entity", "block", "--delete"]);

        const blockDailyStats = getDailyStatsByDateRange(
          await prisma.blockDailyStats.findMany(),
          {
            fromDate,
          }
        );

        expect(blockDailyStats).toEqual([]);
      });

      it("should delete data from the earliest date when no start date is provided", async () => {
        const toDate = "2023-08-05";

        await daily(["--to", toDate, "--entity", "block", "--delete"]);

        const blockDailyStats = getDailyStatsByDateRange(
          await prisma.blockDailyStats.findMany(),
          {
            toDate,
          }
        );

        expect(blockDailyStats).toEqual([]);
      });

      it("should fail when starting date is after ending date", async () => {
        await expect(
          daily(["--from", "2021-01-02", "--to", "2021-01-01", "--delete"])
        ).rejects.toThrowErrorMatchingInlineSnapshot(
          '"Daily stats deletion failed: Invalid date period. Start date is after end date"'
        );
      });
    });
  });
});
