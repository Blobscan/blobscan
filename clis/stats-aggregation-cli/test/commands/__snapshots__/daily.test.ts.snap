// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Daily command > when aggregating daily stats > should aggregate all entities' data when no entity is provided 1`] = `
[
  {
    "avgBlobSize": 1066.666666666667,
    "day": 2022-10-16T00:00:00.000Z,
    "totalBlobSize": 6400n,
    "totalBlobs": 6,
    "totalUniqueBlobs": 3,
  },
  {
    "avgBlobSize": 1250,
    "day": 2023-05-10T00:00:00.000Z,
    "totalBlobSize": 7500n,
    "totalBlobs": 6,
    "totalUniqueBlobs": 6,
  },
  {
    "avgBlobSize": 1033.333333333333,
    "day": 2023-08-03T00:00:00.000Z,
    "totalBlobSize": 6200n,
    "totalBlobs": 6,
    "totalUniqueBlobs": 2,
  },
  {
    "avgBlobSize": 1180,
    "day": 2023-08-20T00:00:00.000Z,
    "totalBlobSize": 5900n,
    "totalBlobs": 5,
    "totalUniqueBlobs": 4,
  },
  {
    "avgBlobSize": 1000,
    "day": 2023-08-28T00:00:00.000Z,
    "totalBlobSize": 4000n,
    "totalBlobs": 4,
    "totalUniqueBlobs": 1,
  },
  {
    "avgBlobSize": 1280,
    "day": 2023-08-31T00:00:00.000Z,
    "totalBlobSize": 6400n,
    "totalBlobs": 5,
    "totalUniqueBlobs": 3,
  },
]
`;

exports[`Daily command > when aggregating daily stats > should aggregate all entities' data when no entity is provided 2`] = `
[
  {
    "avgBlobAsCalldataFee": 44941160,
    "avgBlobFee": 17301504,
    "avgBlobGasPrice": 22,
    "day": 2022-10-16T00:00:00.000Z,
    "totalBlobAsCalldataFee": "44941160",
    "totalBlobAsCalldataGasUsed": "2042780",
    "totalBlobFee": "17301504",
    "totalBlobGasUsed": "786432",
    "totalBlocks": 1,
  },
  {
    "avgBlobAsCalldataFee": 5610000,
    "avgBlobFee": 121000000,
    "avgBlobGasPrice": 22,
    "day": 2023-05-10T00:00:00.000Z,
    "totalBlobAsCalldataFee": "5610000",
    "totalBlobAsCalldataGasUsed": "255000",
    "totalBlobFee": "121000000",
    "totalBlobGasUsed": "5500000",
    "totalBlocks": 1,
  },
  {
    "avgBlobAsCalldataFee": 5610000,
    "avgBlobFee": 121000000,
    "avgBlobGasPrice": 22,
    "day": 2023-08-03T00:00:00.000Z,
    "totalBlobAsCalldataFee": "5610000",
    "totalBlobAsCalldataGasUsed": "255000",
    "totalBlobFee": "121000000",
    "totalBlobGasUsed": "5500000",
    "totalBlocks": 1,
  },
  {
    "avgBlobAsCalldataFee": 5610000,
    "avgBlobFee": 121000000,
    "avgBlobGasPrice": 22,
    "day": 2023-08-20T00:00:00.000Z,
    "totalBlobAsCalldataFee": "5610000",
    "totalBlobAsCalldataGasUsed": "255000",
    "totalBlobFee": "121000000",
    "totalBlobGasUsed": "5500000",
    "totalBlocks": 1,
  },
  {
    "avgBlobAsCalldataFee": 5000000,
    "avgBlobFee": 100000000,
    "avgBlobGasPrice": 20,
    "day": 2023-08-28T00:00:00.000Z,
    "totalBlobAsCalldataFee": "5000000",
    "totalBlobAsCalldataGasUsed": "250000",
    "totalBlobFee": "100000000",
    "totalBlobGasUsed": "5000000",
    "totalBlocks": 1,
  },
  {
    "avgBlobAsCalldataFee": 5610000,
    "avgBlobFee": 121000000,
    "avgBlobGasPrice": 22,
    "day": 2023-08-31T00:00:00.000Z,
    "totalBlobAsCalldataFee": "16830000",
    "totalBlobAsCalldataGasUsed": "765000",
    "totalBlobFee": "363000000",
    "totalBlobGasUsed": "16500000",
    "totalBlocks": 3,
  },
]
`;

exports[`Daily command > when aggregating daily stats > should aggregate all entities' data when no entity is provided 3`] = `
[
  {
    "avgMaxBlobGasFee": 110,
    "day": 2022-10-16T00:00:00.000Z,
    "totalTransactions": 3,
    "totalUniqueReceivers": 3,
    "totalUniqueSenders": 3,
  },
  {
    "avgMaxBlobGasFee": 100,
    "day": 2023-05-10T00:00:00.000Z,
    "totalTransactions": 1,
    "totalUniqueReceivers": 1,
    "totalUniqueSenders": 1,
  },
  {
    "avgMaxBlobGasFee": 100,
    "day": 2023-08-03T00:00:00.000Z,
    "totalTransactions": 2,
    "totalUniqueReceivers": 1,
    "totalUniqueSenders": 1,
  },
  {
    "avgMaxBlobGasFee": 100,
    "day": 2023-08-20T00:00:00.000Z,
    "totalTransactions": 5,
    "totalUniqueReceivers": 3,
    "totalUniqueSenders": 4,
  },
  {
    "avgMaxBlobGasFee": 100,
    "day": 2023-08-28T00:00:00.000Z,
    "totalTransactions": 2,
    "totalUniqueReceivers": 1,
    "totalUniqueSenders": 2,
  },
  {
    "avgMaxBlobGasFee": 100,
    "day": 2023-08-31T00:00:00.000Z,
    "totalTransactions": 3,
    "totalUniqueReceivers": 2,
    "totalUniqueSenders": 2,
  },
]
`;

exports[`Daily command > when aggregating daily stats > should aggregate blob data when the blob entity is provided 1`] = `
[
  {
    "avgBlobSize": 1066.666666666667,
    "day": 2022-10-16T00:00:00.000Z,
    "totalBlobSize": 6400n,
    "totalBlobs": 6,
    "totalUniqueBlobs": 3,
  },
  {
    "avgBlobSize": 1250,
    "day": 2023-05-10T00:00:00.000Z,
    "totalBlobSize": 7500n,
    "totalBlobs": 6,
    "totalUniqueBlobs": 6,
  },
  {
    "avgBlobSize": 1033.333333333333,
    "day": 2023-08-03T00:00:00.000Z,
    "totalBlobSize": 6200n,
    "totalBlobs": 6,
    "totalUniqueBlobs": 2,
  },
  {
    "avgBlobSize": 1180,
    "day": 2023-08-20T00:00:00.000Z,
    "totalBlobSize": 5900n,
    "totalBlobs": 5,
    "totalUniqueBlobs": 4,
  },
  {
    "avgBlobSize": 1000,
    "day": 2023-08-28T00:00:00.000Z,
    "totalBlobSize": 4000n,
    "totalBlobs": 4,
    "totalUniqueBlobs": 1,
  },
  {
    "avgBlobSize": 1280,
    "day": 2023-08-31T00:00:00.000Z,
    "totalBlobSize": 6400n,
    "totalBlobs": 5,
    "totalUniqueBlobs": 3,
  },
]
`;
